cachi2/core/models/validators.py

```diff
@@ -41,3 +44,40 @@ def check_sane_relpath(path: Path) -> Path:
     if os.path.pardir in path.parts:
         raise ValueError(f"path contains {os.path.pardir}: {path}")
     return path
+
+
+def _normalize_pip_versions(version: str) -> str:
+    """Normalize Pip versions that only contain a major and a minor number.
+
+    Pip might ommit the patch version number if it's 0 (e.g. '2.0')
+    """
+    major_minor_only_regex = re.compile(r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)$")
+
+    if major_minor_only_regex.match(version):
+        return version + ".0"
+
+    return version
+
+
+def _normalize_gomod_versions(version: str) -> str:
+    """Remove the 'v' prefix from a gomod version."""
+    gomod_semver_prefix = "v"
+
+    if len(version) > 0 and version[0] == gomod_semver_prefix:
+        return version[1:]
+
+    return version
+
+
+def validate_semver(version: str) -> str:
+    """Validate if a string is a valid semver.
+
+    Small variations that happen for gomod and pip versions are considered valid.
+    """
+    normalized_version = _normalize_gomod_versions(version)
+    normalized_version = _normalize_pip_versions(normalized_version)
+
+    # Will throw a ValueError in case the string is not a valid semver
+    semver.parse(normalized_version)
+
+    return version
```

Comment:
Apologies for the misleading comment about semver earlier - what I meant is that we should only report versions if they follow the specific package manager's semver-ish versioning scheme. Those are not always compatible with semver and I don't think the normalization being done here is sufficient (take a look at [PEP 440](https://peps.python.org/pep-0440/) if you dare).

For python, if the version is something you would find on PyPI, report it as the version. if the "version" is `https://...` or `git+...`, don't report it as the version.

For gomod, always report the version unless it's a file path.

I really don't think this should be done on the SBOM level, but on the package manager level.

Reply:
Thanks for linking the PEP 440 doc. Yeah, I completely agree that it should be done in the package manager level. 

This is just a temporary implementation, I was thinking we could for now just omit versions that are not semver-ish, and improve the package manager code to handle correct version outputting (from a CycloneDx perspective) and deduplication in a follow-up PR. Wdyt?

For now, I think I will with an approach as simple as omitting versions that start with `https` or `git`. That should be enough, right?

Reply:
Ah, ok. Yeah, we can probably just omit `http(s)?://`, `git+`, `./` and `../` for now